### PR TITLE
Monthly Column Statements

### PR DESCRIPTION
**Pull request by sampoder**
## Summary of the problem

Requested by Bence.

## Describe your changes

@garyhtou I'd like your thoughts on this before I move forward, here's the CSV it's generator:

Screenshot 2024-01-25 at 9 10 22 PM

Questions that have are: do we want to store the CSV as a string or as a file?

Still need to create the model and stuff.

### CODE REVIEW

**Comment by garyhtou** [comment]
As a file!

File: app/models/column_statement.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Table name: column_statements
+#
+#  id         :bigint           not null, primary key
+#  end_date   :datetime
+#  start_date :datetime
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+class ColumnStatement < ApplicationRecord
```
Comment (garyhtou) [inline app/models/column_statement.rb]: Please namespace as `Column::Statement` (`models/column/statement.rb`). Thank you!

File: app/jobs/column_reports_job.rb [hunk 2]
```diff
@@ -1,14 +1,68 @@
 # frozen_string_literal: true
 
+require "csv"
+
 class ColumnReportsJob < ApplicationJob
```
Comment (garyhtou) [inline app/jobs/column_reports_job.rb]: Let's call this `Column::StatementJob` (file path will be `jobs/column/statement_job.rb`)

File: app/jobs/column_reports_job.rb [hunk 3]
```diff
@@ -1,14 +1,68 @@
...
+    end
+
+    column_statement = ColumnStatement.new
+    column_statement.file.attach(io: File.open(file), filename: "column_statement_report_#{end_date.iso8601}.csv")
+    column_statement.start_date = start_date
+    column_statement.end_date = end_date
+    column_statement.save!
+
+    file.close
+    file.unlink
```
Comment (garyhtou) [inline app/jobs/column_reports_job.rb]: Using the block format will automatically close and unlink the file

https://ruby-doc.org/stdlib-2.5.6/libdoc/tempfile/rdoc/Tempfile.html#method-c-create

```suggestion
    Tempfile.create("column_statement_csv") do |file|
      CSV.open(file, "w", headers: @headers, write_headers: true) do |csv|
        rows.each { |row| csv << row }
      end

      column_statement = ColumnStatement.new
      column_statement.file.attach(io: File.open(file), filename: "column_statement_report_#{end_date.iso8601}.csv")
      column_statement.start_date = start_date
      column_statement.end_date = end_date
      column_statement.save!
    end
```

File: app/jobs/column_reports_job.rb [hunk 4]
```diff
@@ -1,14 +1,68 @@
 # frozen_string_literal: true
 
+require "csv"
```
Comment (garyhtou) [inline app/jobs/column_reports_job.rb]: shouldn't be necessary

Comment (sampoder) [reply app/jobs/column_reports_job.rb]:
```suggestion
```

File: app/models/user.rb [hunk 5]
```diff
@@ -194,7 +194,7 @@ def admin_dropdown_description
   end
 
   def birthday?
-    birthday.present? && birthday.month == Date.today.month && birthday.day == Date.today.day
+    true
```
Comment (garyhtou) [inline app/models/user.rb]: 🎉HAPPY BIRTHDAY!!🎂

unrelated? lol

**Review by garyhtou — COMMENTED** [review]
Do we want a monthly job so that it will automatically create a `Column::Statement` on the first of each month?

File: app/models/user.rb (hunk 5 above)
Comment (sampoder) [reply app/models/user.rb]: lol yeah
```suggestion
    birthday.present? && birthday.month == Date.today.month && birthday.day == Date.today.day
```

**Comment by sampoder** [comment]
This job should do that: https://github.com/hackclub/hcb/blob/a4faf9a8ba40f66b1d95c37a0cb4a8f39deaf793/app/jobs/column/reports_job.rb.

Let me quickly update the schedule

File: app/views/admin/column_statements.html.erb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      <th class="w-32">End Date</th>
+      <th>Statement</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @statements.each do |statement| %>
+      <tr>
+        <td><%= statement.start_date&.strftime("%B, %Y") %></td>
+        <td><%= statement.start_date&.strftime("%Y-%m-%d") %></td>
+        <td><%= statement.end_date&.strftime("%Y-%m-%d") %></td>
```
Comment (garyhtou) [inline app/views/admin/column_statements.html.erb]: is there a case where these are `nil`?

File: app/jobs/column/reports_job.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Column
+  class ReportsJob < ApplicationJob
+    queue_as :low
+    def perform(start_date: Date.today.prev_month.beginning_of_month, end_date: Date.today.prev_month.end_of_month)
+
```
Comment (garyhtou) [inline app/jobs/column/reports_job.rb]: This helps ensure all our reports are actually at start and end of month
```suggestion
    def perform(date = Date.today.prev_month)
      start_date = date.beginning_of_month
      end_date = date.end_of_month

```

**Review by garyhtou — APPROVED** [review]

**Review by garyhtou — COMMENTED** [review]
Great! the only thing we're missing here is
> Balance at the start and end of the month
>https://github.com/hackclub/hcb/issues/5057

File: app/jobs/column/statement_job.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                           return "#{ach_transfer["company_name"]} #{ach_transfer["company_entry_description"]}"
+                         elsif transaction_id.start_with? "book"
+                           book_transfer = ColumnService.get "/transfers/book/#{transaction_id}"
+
+                           return book_transfer["description"]
+                         end
+                         "TRANSACTION"
+                       }],
+        [:amount_cents, ->(t) { t["available_amount"] }],
+        [:check_number, ->(t) { "" }]
```
Comment (garyhtou) [inline app/jobs/column/statement_job.rb]: Can we add in the `check_number`?

File: app/jobs/column/statement_job.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+      template = [
+        # Must be wrapped in lambdas
+        [:date_posted, ->(t) { t["effective_at"] }],
+        [:description, ->(t) {
+                         transaction_id = t["transaction_id"]
+                         if transaction_id.start_with? "acht" # TODO: use `transaction_type` instead
+                           ach_transfer = ColumnService.ach_transfer(transaction_id)
+                           return "#{ach_transfer["company_name"]} #{ach_transfer["company_entry_description"]}"
+                         elsif transaction_id.start_with? "book"
```
Comment (garyhtou) [inline app/jobs/column/statement_job.rb]: chkt?

Comment (sampoder) [reply app/jobs/column/statement_job.rb]:
```suggestion
                         elsif transaction_id.start_with? "chkt"
```

**Review by garyhtou — APPROVED** [review]

**Comment by garyhtou** [comment]
@sampoder can you backfill October 2023 until January 2024 via the console?

FYI, October 2023 might be empty

**Pull request merged**